content/articles/firebase-offline-caching/index.md

```diff
@@ -0,0 +1,294 @@
+    implementation 'com.google.firebase:firebase-database-ktx'
+
+    //Glide
+    implementation 'com.github.bumptech.glide:glide:4.13.0'
+    annotationProcessor 'com.github.bumptech.glide:compiler:4.13.0'
+
+    // ViewModel
+    implementation "androidx.lifecycle:lifecycle-viewmodel-ktx:$lifecycle_version"
+
+    // Livedata
+    implementation "androidx.lifecycle:lifecycle-livedata-ktx:$lifecycle_version"
+
+    // Coroutines
+    implementation "org.jetbrains.kotlinx:kotlinx-coroutines-android:$coroutines_version"
+    implementation 'org.jetbrains.kotlinx:kotlinx-coroutines-play-services:1.5.1'
+
+    // Room
+    implementation "androidx.room:room-runtime:2.4.2"
+    kapt "androidx.room:room-compiler:2.4.2"
+```
+
+The app we are going to build will have many classes, to avoid confusion and maintain clean code, we will create packages that will separate the classes. Right-click on the main/root directly and create the following packages; `ui`, `util`, `repository`, `viewmodel`, `adapter`, `model`, `data`(`local`). Data has a sub-package for holding the local database.
+
+The project build will have a single screen that has a `RecyclerView` for displaying the data fetched from the real-time database a `Progressbar` and a `SearchView`. See the project design below:
+
+![screen layout](/engineering-education/firebase-offline-caching/screen-layout.jpeg)
+
+Now that we have set the project structure, we can dive into implementation.
+
+#### Step 1 - Creating model class
+The model class is a data class whose main purpose is to hold the data that is being fetched from the Firebase real-time database. Since we want to cache the data using Room, we must annotate the model class with the `@Entity` for this class to serve both as the remote and local data class. 
+
+Optionally, you can as well create a separate model data class and annotate with `@Entity`. This model class will hold the data from [this](https://drive.google.com/file/d/1j_Oyh3Yce2WR5JqB1wPtUF1UMZePbN0l/view?usp=sharing) JSON file. You should import this file into your Firebase real-time database.
+
+```kotlin
+@Entity(tableName = "items")
+data class ItemsEntity (
+    @PrimaryKey(autoGenerate = true) val id: Int?,
+    val itemImage: String? = null,
+    val itemName: String? = null,
+    val itemOldPrice: String? = null,
+    val itemPrice: String? = null
+)
+```
+
+#### Step 2 - Defining utility items
+In the `util` package, define the following utility items. The utility package is for defining utility methods constants that can be reused in various classes.
+
+```kotlin
+
+```
+This sealed class will help us in handling the three states of our real-time database network calls i.e `Loading`, `Success`, and `Error`.
+
+#### Step 3 - Creating data classes
+We will define our `DAO` and `Database` classes in the `data` package.
+
+#### Creating DAO interface
+DAO is an interface that has methods for manipulating the local database. It should look like the one below:
+
+@Database(
```

Comment:
Add backticks and use the proper language/code identifier. Ensure all your code runs perfectly.